apps/aehttp/src/sc_ws_handler.erl

```diff
@@ -52,46 +85,125 @@ websocket_init(Params) ->
             {stop, undefined}
     end.
 
+reconnect_to_fsm_(#{ channel_id := ChanId
+                   , role       := Role
+                   , pub_key    := Pubkey
+                   , signed_tx  := SignedTx } = Opts, Handler, OnError) ->
+    case aesc_fsm:where(ChanId, Role) of
+        undefined ->
+            lager:debug("where(~p, ~p) -> undefined", [ChanId, Role]),
+            maybe_reestablish(Opts, Handler, OnError);
+        #{ fsm_pid := Fsm, pub_key := Pubkey } ->
+            %% At this point, we haven't verified the signature.
+            %% This is done by the fsm.
+            lager:debug("Found FSM = ~p", [Fsm]),
+            case aesc_fsm:reconnect_client(Fsm, self(), SignedTx) of
+                ok ->
+                    MRef = erlang:monitor(process, Fsm),
+                    { ok, Handler#handler{ fsm_pid  = Fsm
+                                         , fsm_mref = MRef } };
+                {error, E} ->
+                    OnError(E)
+            end
+    end.
+
+%% The fsm is not running. Perhaps there is cached state, warranting a reestablish?
+%%
+maybe_reestablish(#{ channel_id := ChId
+                   , pub_key    := Pubkey } = ReconnectOpts, Handler0, OnError) ->
+    lager:debug("ReconnectOpts = ~p",
+                [aesc_utils:censor_init_opts(ReconnectOpts)]),
+    case maps:find(state_password, ReconnectOpts) of
+        {ok, Password} ->
+            case aesc_state_cache:reestablish(ChId, Pubkey, Password) of
+                {ok, State, Opts} ->
+                    lager:debug("Fetched state. Opts = ~p", [Opts]),
+                    Handler = update_handler_for_reestablish(Handler0, Opts),
+                    try aesc_offchain_state:get_latest_signed_tx(State) of
+                        {_, SignedTx} ->
+                            lager:debug("Latest tx = ~p", [SignedTx]),
+                            ExpandedOpts = expand_cached_opts(Opts),
+                            websocket_init_parsed(
+                              Handler,
+                              ExpandedOpts#{ existing_channel_id => ChId
+                                           , offchain_tx => SignedTx })
+                    catch
+                        error:_E ->
+                            lager:debug("Failed getting latest signed tx: ~p", [_E]),
+                            OnError(not_found)
+                    end;
+                {error, _} ->
+                    lager:debug("No cached state", []),
+                    OnError(not_found)
+            end;
+        error ->
+            lager:debug("No state_password specified", []),
+            OnError(not_found)
+    end.
+
+check_existing_client(Client, Opts, Handler) ->
+    OnError = fun(E) ->
+                      handler_init_error(E, Handler)
+              end,
+    T = time_since_last_dispatch(Client),
+    lager:debug("Time since last dispatch (~p): ~p", [Client, T]),
+    if is_integer(T), T < 5000 ->
```

Comment:
The timeout should be documented and configurable. The configuration can be in a different PR though.

Reply:
This PR has enough stuff already...